### PR TITLE
speed up chunking & add separator chunking

### PR DESCRIPTION
**Pull request by rangehow**
Hey, so this PR's got two main changes:

1. We're now converting docs to tokens in bulk, which is giving us a sweet 3X speed boost when dealing with a ton of docs (we tested it with 30k) by chunking_by_token_size. It's not gonna make much difference for small-scale stuff, but 30k is still pretty much toy-level (both industry and research usually work with way more). So yeah, this is definitely a solid upgrade.

2. We've added support for separator-based splitting without needing any extra dependencies. This splitting method tries to keep the grammar structure intact, meaning you'll always get complete clauses or sentences (if without any overlap). We tweaked the logic from langchain, so it might not be exactly the same, but it does what it says on the tin.

### CODE REVIEW

File: nano_graphrag/_spliter.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            else:
+                overlap_tokens = chunks[i-1][-self._chunk_overlap:]
+                new_chunk = overlap_tokens + chunk
+                if self._length_function(new_chunk) > self._chunk_size:
+                    new_chunk = new_chunk[-self._chunk_size:]
+                new_chunks.append(new_chunk)
+        return new_chunks
+
+# EXAMPLE USAGE
+if __name__ == "__main__":
```
Comment (gusye1234) [inline nano_graphrag/_spliter.py:92]: May be move this section to `tests`? Make sure the splitter is tested

File: nano_graphrag/graphrag.py [hunk 2]
```diff
@@ -264,20 +266,22 @@ async def ainsert(self, string_or_strings):
...
+            
+            new_docs_list=list(new_docs.items())
+            docs=[new_doc[1]["content"] for new_doc in new_docs_list]
+            doc_keys=[new_doc[0] for new_doc in new_docs_list]
+            
+            
+            ENCODER = tiktoken.encoding_for_model("gpt-4o")
+            tokens=ENCODER.encode_batch(docs,num_threads=16)
+            chunks=self.chunk_func(tokens,overlap_token_size=self.chunk_overlap_token_size,
+                        max_token_size=self.chunk_token_size,doc_keys=doc_keys,tiktoken_model=ENCODER)
```
Comment (gusye1234) [inline nano_graphrag/graphrag.py:280]: Can we abstract this block into a function in `_op`? Make insert process clean and short

Comment (rangehow) [reply nano_graphrag/graphrag.py:280]: I don't recommend adding another layer of abstraction here. While this unavoidable code does look a bit tedious, it doesn't significantly increase the line count. If we were to further abstract this section, we'd likely need to name this entire behavior. It almost represents a complete phase（chunking） and is hard to consider as only a single operator (_op).  , I'm still open to hearing your thoughts on this. But I might need some more detailed ideas.

File: nano_graphrag/_spliter.py (hunk 1 above)
Comment (rangehow) [reply nano_graphrag/_spliter.py:92]: OK. Let me try this.

File: nano_graphrag/graphrag.py (hunk 2 above)
Comment (gusye1234) [reply nano_graphrag/graphrag.py:280]: Maybe we can change the abstract of chunk_func? 
You can directly pass the `new_docs` to chunk_func and return the same things as now. So the parameters of chunk func is much more clean, so as the main process of insert .

File: nano_graphrag/_spliter.py (hunk 1 above)
Comment (rangehow) [reply nano_graphrag/_spliter.py]: done

Comment (rangehow) [reply nano_graphrag/_spliter.py]: oh god wait

Comment (rangehow) [reply nano_graphrag/_spliter.py]: done

**Comment by rangehow** [comment]
NOW I GUESS ALL SHOULD BE WELL : )

File: nano_graphrag/_op.py [hunk 3]
```diff
@@ -28,27 +30,102 @@
...
+
+
 def chunking_by_token_size(
-    content: str, overlap_token_size=128, max_token_size=1024, tiktoken_model="gpt-4o"
-):
-    tokens = encode_string_by_tiktoken(content, model_name=tiktoken_model)
-    results = []
-    for index, start in enumerate(
-        range(0, len(tokens), max_token_size - overlap_token_size)
+        tokens_list: list[int], doc_keys,tiktoken_model, overlap_token_size=128, max_token_size=1024,
```
Comment (gusye1234) [inline nano_graphrag/_op.py:36]: Shouldn't `tokens_list` be `list[list[int]]`?

File: nano_graphrag/_op.py [hunk 4]
```diff
@@ -28,27 +30,102 @@
...
+                        "tokens": lengths[i],
+                        "content": chunk.strip(),
+                        "chunk_order_index": i,
+                        "full_doc_id":doc_keys[index],
+                    }
+                )
+
+        return results
+
+def chunking_by_seperators(tokens_list: list[int], doc_keys,tiktoken_model, overlap_token_size=128, max_token_size=1024 ):
```
Comment (gusye1234) [inline nano_graphrag/_op.py:63]: Shouldn't tokens_list be list[list[int]]?

File: nano_graphrag/graphrag.py [hunk 5]
```diff
@@ -65,7 +68,7 @@ class GraphRAG:
     enable_naive_rag: bool = False
 
     # text chunking
-    chunk_func: Callable[[str, Optional[int], Optional[int], Optional[str]], List[Dict[str, Union[str, int]]]] = chunking_by_token_size
+    chunk_func: Callable[[str,List[str],tiktoken.Encoding, Optional[int], Optional[int], ], List[Dict[str, Union[str, int]]]] = chunking_by_token_size
```
Comment (gusye1234) [inline nano_graphrag/graphrag.py:71]: The first parameter of `chunk_func` should be `list[list[int]]`, right?

**Review by gusye1234 — COMMENTED** [review]
Great works! Few typing errors I think

File: nano_graphrag/_op.py (hunk 3 above)
Comment (rangehow) [reply nano_graphrag/_op.py:36]: DONE

**Pull request merged**